Add memory support to IonQJob and update tests for error handling

splch opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short, detailed and understandable for all.
⚠️ If your pull request addresses an open issue, please link to the issue.

✅ I have added tests to cover my changes.
✅ I have updated the documentation accordingly.
✅ I have read the CONTRIBUTING document.

😊 Thank you for helping make this project better!
-->

### Summary

adds memory to support shotwise output

### Details and comments

- need to get shotwise endpoint only when requested
- do not attempt shotwise for ideal simulations

splch committed:
Add memory support to IonQJob and update tests for error handling

splch committed:
Reformat warning filters in pytest section for improved readability

splch committed:
Add warning filter for IonQ backend optimization level in tox.ini

splch committed:
Refactor IonQJob to handle multiple result URLs and improve results retrieval logic

splch committed:
Update type hints for optional parameters and increment version to 1.0.3

splch committed:
Refactor type hints in IonQClient to use union types for optional parameters

splch committed:
Add memory retrieval and histogram support in IonQJob and tests

splch committed:
Fix linting failures

splch reviewed (commented):
comment below vv

splch commented on qiskit_ionq/ionq_job.py:
```diff
@@ -543,16 +602,31 @@ def _format_result(self, data):
+                        width=(qiskit_header[i] or {}).get(
+                            "memory_slots", len(self._clbits[i])
+                        ),
+                    )
```
i dont like how this looks

splch committed:
Add memory fetching support and tests for IonQJob

splch committed:
Merge branch 'main' into add-memory-support

Resolve conflicts with #235, #236, #238, #239, #240, #241, #246, #251.

Carry forward (review feedback from probing the live API):

- Source the shotwise URL from the server response. _format_result()
  populates self._results_urls from response['results'] (now defensive
  against results: null), and _fetch_memory() uses
  self._results_urls.get('shots') instead of a hardcoded /v0.4 path.
  Empirically confirmed against api.ionq.co/v0.4: top-level results
  contain {'probabilities', 'histogram', 'shots'} for noisy-sim and
  QPU jobs once shotwise rollout is in place, and absent ('results':
  null) for dry-run jobs.
- Replace get_memory()'s circuit.name access with a getattr-with-
  fallback so passing a circuit index (int) or name (str) raises
  IonQBackendError with a useful label instead of AttributeError.
- Fix _build_memory()'s memory_slots fallback for clbits=None.

Tests:

- Update test_backend_memory to feed the real wire format ('0', '1',
  '2', '3') instead of 2-char bitstring lookalikes that only coincide
  for 2-qubit outputs.
- Add test_build_memory_3q_wire_format to lock the wire-format
  contract (decimal-encoded outcome integers, not pre-formatted
  bitstrings) using 3-qubit fixtures.
- Add test_no_shots_url_returns_none: jobs whose response omits the
  shots URL must skip the fetch cleanly without HTTP round-trip.
- Update test_dry_run_no_results_url to assert on the new plural
  self._results_urls attribute (was self._results_url).

splch committed:
Shorten test name to satisfy pylint function-rgx (≤31 chars)

CI's lint job runs pylint on test/ too, where .pylintrc pins
function-rgx=[a-z_][a-z0-9_]{2,30}$ (max 31 chars).
test_build_memory_3q_wire_format was 32.

splch committed:
Align shotwise with v0.4 OpenAPI schema; guard against shape drift

Cross-checked PR #234's shotwise implementation against api.ionq.co/v0.4
and ionq-core 0.1.1 (the official low-level client):

URL pattern
  Spec defines CircuitJobResult.shots.url as a top-level URL slot but
  does not document a corresponding GET operation; ionq-core only
  exposes the per-variant /jobs/<id>/variants/<vid>/results/shots
  (get_variant_shots). PR uses the response-supplied url from the
  CircuitJobResult.shots.url slot, which is the correct abstraction
  for Qiskit's flat per-shot 'memory' list (the per-variant URL would
  require client-side aggregation across N debiasing variants). This
  is aligned with the schema's structural intent and survives any
  future change in the underlying URL.

Wire-format
  Spec types GetVariantResultsResponse as 'additionalProperties:
  number/double' (i.e. dict-of-floats). Live API instead serves shots
  as an ordered list of decimal-encoded outcome integers (verified
  empirically: 'Content-Type: application/json', body opens with
  ['3','3','0',...]). ionq-core's generated parser actually crashes on
  the real response (ValueError in dict() of a list-of-strings) - so
  qiskit-ionq can't delegate to it here.

  PR matches the live wire format. To prevent silent corruption if the
  API ever flips to the spec-documented dict shape (which would lose
  shot order), _build_memory now raises IonQJobError on non-list
  inputs with a message that points at the relevant schema slot and
  the qiskit-ionq issue tracker.

  Also tightened the docstrings on _build_memory and _fetch_memory to
  cite the schema field name explicitly so future maintainers can
  trace back to the spec.

splch committed:
Trim verbose comments, docstrings, and the dict-shape guard

Removes ~120 lines of explanatory prose and over-defensive code added in
21890ff. Keeps the structural behavior - shots URL sourced from
CircuitJobResult.shots.url, list-shaped wire format, clean skip when
absent - and drops:

  - the isinstance(raw_shots, (list, tuple)) guard plus its test (API
    contract is settled; matches existing repo style which trusts the
    wire shape from _build_counts onward),
  - multi-paragraph docstrings on _build_memory / _fetch_memory /
    get_memory (one short paragraph each, mentioning the schema slot
    by name only where it earns its keep),
  - mid-function comments restating what the code does,
  - test docstrings re-explaining the wire format already locked by
    test_build_memory_3q_format,
  - the verbose README paragraph (trimmed to match the surrounding
    sibling sections).

splch committed:
Address review: hoist shots fetch, skip multi-circuit, surface fetch failures

Review feedback on PR #234 (https://github.com/qiskit-community/qiskit-ionq/pull/234):

- Hoist the /results/shots fetch out of the per-circuit loop. Previously
  _fetch_memory ran inside 'for i in range(self._num_circuits)' and would
  have fired N redundant HTTP requests for multi-circuit jobs, each
  reusing the same flat shot list and assigning it (with each circuit's
  own clbit map) to every result slot.

- Decline multi-circuit memory rather than mis-assign. The v0.4
  CircuitJobResult.shots.url slot is a single top-level URL with no
  documented per-circuit split (only per-variant URLs exist via
  get_variant_shots). When _num_circuits > 1, warn and emit memory=None
  for every result. Covered by test_multi_circuit_skips_memory.

- Surface IonQAPIError on the shots fetch as a UserWarning instead of
  silent memory=None. Users who asked for memory and got None can now
  see why. Covered by test_shots_fetch_warns_on_error.

- Add test_get_memory_raises_when_off for the existing IonQBackendError
  raise-path on jobs submitted with memory=False (previously untested).

- Drop the redundant isinstance(header, dict) guard in _fetch_memory's
  successor; the only caller passes header = qiskit_header[i] or {}.

- Rename test_dry_run_no_results_url -> test_dry_run_no_results_urls to
  match the attribute rename in this branch (_results_url -> _results_urls).

- Add pytest.warns wrapper to test_multi_null_meta_result so the new
  multi-circuit warning is asserted rather than leaking to the summary.

splch committed:
Default memory=False to match qiskit/qiskit-aer backend.run convention

Following the discussion of which default makes more sense, flip the
`memory` default from True to False. Rationale:

- The legacy qiskit.execute() documented memory as 'Default: False'
- IBMQBackend.run() docs and examples uniformly show 'set memory=True'
- qiskit-aer AerSimulator.run() tutorial says 'enabled by setting
  memory=True in the run'
- qiskit-braket-provider and other vendor backends follow the same
  opt-in convention
- Every Stack Exchange answer about per-shot results since 2019 starts
  with 'you need to use memory=True'

With the previous default-on behavior every QPU/noisy-sim job paid an
extra GET /results/shots round trip plus a JSON body of N
decimal-encoded outcome integers (~5-50 KB at typical shot counts), even
when the user only wanted counts. The V2-primitive argument for
default-on (per-shot data is implicit in SamplerV2 results) doesn't
translate here because qiskit-ionq's memory comes from a separate
endpoint, not the same response.

Source:

- IonQBackend._default_options(): memory=True -> False
- IonQJob.__init__: both passed_args and bare default flipped to False
- README: explain opt-in and call out the convention

Tests:

- expected_result (test_result family) reverts to its pre-PR shape
  (Result.to_dict() omits memory: None during serialization)
- Drop now-unused /results/shots mocks from test_result variants,
  test_counts__simulator_probs, test_counts_and_probs_from_job,
  test_result__meas_mapped, test_counts_marginalize, and the
  formatted_result conftest fixture
- test_backend_memory, test_multi_circuit_skips_memory,
  test_shots_fetch_warns_on_error, test_no_shots_url_returns_none,
  test_multi_null_meta_result: explicitly construct the job with
  passed_args={memory: True, ...} since they assert behaviour that
  requires memory to be opted in
- test_no_memory_skips_shots and test_get_memory_raises_when_off
  already passed memory=False explicitly, so no change

splch commented:
We standardized cross-SDK on `memory` + off by default (this PR ✅). One ask before merge: add multi-circuit support. Each child job advertises its own `results.shots.url` — iterate `child_job_ids` and assemble a per-circuit `memory` list. Same pattern as PennyLane-IonQ#178's `_retrieve_child_job_shots`. This unblocks #237 properly instead of forcing it down the resample path.

guenp commented on qiskit_ionq/ionq_job.py:
```diff
@@ -144,6 +145,33 @@ def _build_counts(  # pylint: disable=too-many-positional-arguments
+    keys); this function applies the ``clbits`` mapping and returns one
+    MSB-left bitstring per shot.
+    """
+    if not clbits:
```
```suggestion
    if clbits is None:
```

I would make this an explicit check, following the Zen of Python

splch replied:
Done — switched to `if clbits is None:` so the check lines up with the `list[int] | None` signature.

guenp commented on qiskit_ionq/ionq_job.py:
```diff
@@ -144,6 +145,33 @@ def _build_counts(  # pylint: disable=too-many-positional-arguments
+        clbits = list(range(n_qubits))
+    out_width = int(width) if width is not None else len(clbits) if clbits else n_qubits
+
+    def remap_one(val: int | str) -> str:
```
```suggestion
    def remap_raw_shot(val: int | str) -> str:
```
naming nit

splch replied:
Done.

guenp commented on qiskit_ionq/ionq_job.py:
```diff
@@ -144,6 +145,33 @@ def _build_counts(  # pylint: disable=too-many-positional-arguments
+        )[::-1]
+        return mapped.rjust(out_width, "0")
+
+    return [remap_one(s) for s in raw_shots]
```
```suggestion
    return [remap_raw_shot(s) for s in raw_shots]
```

splch replied:
Done (paired with the helper rename above).

guenp commented on qiskit_ionq/ionq_job.py:
```diff
@@ -291,6 +321,22 @@ def get_counts(self, circuit: Optional[QuantumCircuit] = None) -> dict:
+        Raises :class:`IonQBackendError` if the job was submitted with
+        ``memory=False``.
+        """
+        if self.memory:
```
```suggestion
        if self.memory is True:
```

just better to be explicit imo (for code readability)

splch replied:
Going to keep `if self.memory:` here. PEP 8 *Programming Recommendations* is explicit about this one:

> Don't compare boolean values to True or False using `==`. Use `if greeting:` instead of `if greeting is True:`.

`self.memory` is initialized as a `bool` in `__init__` (`passed_args.pop("memory", False)`), so the truthy form is the idiomatic Python check. The identity comparison would also silently break if someone ever flowed a truthy non-`True` value through that field.

guenp commented on qiskit_ionq/ionq_job.py:
```diff
@@ -430,18 +476,14 @@ def status(self, detailed: bool = False) -> jobstatus.JobStatus | dict:
-                    else _results_url.get("probabilities", {}).get("url")
-                )
+                results = response.get("results") or {}
+                self._results_urls = {
```
Does this satisfy the multi-circuit support mentioned in the most recent comment?

splch replied:
No — that block just records `shots` and `probabilities` as separate URL slots so the rest of the code can pick one. The actual multi-circuit work is in the new commit on top.

`_raw_shots_per_circuit` now iterates `child_job_ids` and fetches each child's own `results.shots.url` (the parent only advertises aggregated probabilities for multi-circuit jobs). Failures degrade per-circuit — one bad child returns `memory=None` for that slot, the others still assemble. Covered by `test_multi_mem_per_child` (happy path) and `test_multi_mem_partial` (one child missing its shots URL).

guenp commented on qiskit_ionq/ionq_job.py:
the docstring doesn't match the function signature, and could use more clarity (i.e. what is the expected schema of `result`)

splch replied:
Rewrote it: the param is `data` (not `result`), and the docstring now spells out the three accepted shapes (`dict[str, float]` single, `dict[str, dict[str, float]]` multi keyed by child id, and the `list[dict[str, float]]` tests use) plus the `int`/`float` accepted for probability values.

guenp commented on qiskit_ionq/ionq_job.py:
```diff
@@ -587,7 +648,8 @@ def _format_result(self, data):
             looks_like_multi = all(
-                isinstance(v, dict) and all(isinstance(p, float) for p in v.values())
+                isinstance(v, dict)
+                and all(isinstance(p, (int, float)) for p in v.values())
```
I'll need a bit more context for this change. Am I inferring correctly that the output format of `data` changed? It looks like it went from `{str: float}` to `{str: (int, float)}`? that seems like a breaking change, and would perhaps require a version bump (if so please note in the PR description). On a related note, should we also check for the old format here for backward compatibility?

splch replied:
Not a format change — strictly a permissiveness bump. Python's `json` module deserializes the JSON literals `0` and `1` as Python `int`, not `float`, so a noiseless-simulator child whose probabilities happen to be exact 0/1 used to fail the heuristic, and `_format_result` silently mis-classified that multi-circuit payload as single-circuit. Old `{str: float}` payloads still match — adding `int` to the isinstance tuple makes the check more permissive, not less. No version bump and no back-compat shim needed; the same `data: dict[str, float]` callers continue to work. The new `_format_result` docstring calls this out so the schema is explicit going forward.

guenp commented on qiskit_ionq/ionq_job.py:
```diff
@@ -610,26 +672,51 @@ def _format_result(self, data):
                 )
+                if raw_shots is not None:
+                    memory_slots = header.get("memory_slots") or len(clbits or [])
+                    memory = _build_memory(raw_shots, n_qubits, clbits, memory_slots)
```
```suggestion
                    memory = _build_memory(raw_shots=raw_shots, n_qubits=n_qubits, clbits=clbits, width=memory_slots)
```

It can be a little risky to skip the variable names, especially if the variable names and the function input names don't match, I would recommend making this explicit

```
_build_memory(
    raw_shots: list[int | str],
    n_qubits: int,
    clbits: list[int] | None,
    width: int | None = None,
)
```

splch replied:
Fixed at the root rather than at the call site: renamed the parameter `width` → `memory_slots` in `_build_memory` so all four positionals match the caller's vocabulary directly. Sprinkling kwargs at the call site would have preserved the impedance mismatch (`memory_slots` → `width`) that prompted the comment; aligning the names is the cleaner read.

guenp commented on qiskit_ionq/ionq_job.py:
```diff
@@ -144,6 +145,33 @@ def _build_counts(  # pylint: disable=too-many-positional-arguments
     return counts, probabilities
 
 
+def _build_memory(
```
```suggestion
def _shots_to_memory_bitstrings(
```
nit: more readable/understandable naming
I think "build" is not really accurate here - we are _converting_ one encoding to another, so imo `a_to_b` would be a better naming paradigm to use

splch replied:
Going to keep `_build_memory`. It has a sibling 30 lines above — `_build_counts` — and the two are the producers `_format_result` assigns into `data.counts` and `data.memory`. Renaming this one alone breaks the established naming family; renaming both is out of scope here (and `_build_counts` does more than convert — it samples from probabilities for the ideal simulator, so "build" actually fits it).

`memory` is also the Qiskit term of art for the list of per-shot bitstrings that lives at `result.data.memory`, so `_build_memory` reads as "build the `memory` field," not "build memory in the OS sense."

guenp reviewed (approved):
lgtm, a few naming and value assertion nits, and curious about the multi-circuit support in the last comment -- it says "before merging", does that mean this is a fix to still be added?

splch committed:
Support per-shot memory for multi-circuit jobs

Multi-circuit jobs previously warned and emitted memory=None because the
parent only advertises aggregated probabilities. Each child job carries
its own results.shots.url, so iterate child_job_ids and fetch each in
turn, indexing into a per-circuit shots list before _build_memory.

_fetch_raw_shots now takes the URL as an argument (was reading from
_results_urls directly) so the same helper handles single-circuit and
per-child fetches with a unified UserWarning on fetch failures. Failures
degrade per-circuit -- one bad child does not poison the whole result.

Replaces test_multi_circuit_skips_memory with two tests covering the
happy path and the partial-failure path; drops memory=True from
test_multi_null_meta_result since null-meta handling is its real focus.

splch committed:
Address review nits in _build_memory and _format_result

- Rename _build_memory's 'width' param to 'memory_slots' so the four
  positionals match the Qiskit-vocabulary names at every call site;
  also rename the inner closure remap_one -> remap_raw_shot.
- Switch the missing-clbits guard from 'if not clbits' to
  'if clbits is None' to match the type signature
  (list[int] | None) explicitly.
- Fix _format_result's docstring: parameter is 'data' (not 'result'),
  and document the three accepted payload shapes plus the int|float
  probability values from noiseless simulators.

No behavior change; per-shot memory for single- and multi-circuit jobs
continues to pass the full suite.

splch commented:
@guenp thanks for the careful pass. Two new commits address the round.

**Multi-circuit memory** (the load-bearing pre-merge ask): `_raw_shots_per_circuit` iterates `child_job_ids` and fetches each child's own `results.shots.url` — the parent only advertises aggregated probabilities for multi-circuit, so the previous "decline rather than mis-assign" branch was leaving real per-shot data on the table. Failures degrade per-circuit; one bad child returns `memory=None` for that slot without poisoning the rest. Covered by `test_multi_mem_per_child` and `test_multi_mem_partial`.

**Review nits accepted:** `if clbits is None` (matches the `list[int] | None` signature), `remap_one` → `remap_raw_shot`, `_format_result` docstring rewritten with the three accepted payload shapes and the `int`/`float` probability values. Also addressed the `_build_memory` call-site mismatch by renaming the parameter `width` → `memory_slots` so all four positionals match the caller's vocabulary directly (cleaner than spreading kwargs at the call site).

**Two declined,** with rationale in the inline threads:
- `if self.memory is True:` — PEP 8 explicitly forbids this form for boolean checks.
- `_build_memory` → `_shots_to_memory_bitstrings` — would break the naming pair with the existing `_build_counts` sibling.

The `(int, float)` change is a permissiveness bump (JSON `0`/`1` deserialize as Python `int`), not a format change — old `{str: float}` payloads still match. No version bump needed; the new docstring documents the schema explicitly.

Full suite (392 tests) green on `pr-234`; pylint clean on both `qiskit_ionq/` and `test/`.
